Amazon Step5 검색자동완성

cocoa-salon opened the pull request:
검색자동완성 기능을 구현하였습니다. 검토 부탁드립니다. 
아직 debounce를 적용하지 않아 추후 수정 작업 시 반영할 예정입니다. 

감사합니다.

Commit:
1.각종 상수 별도 관리

-함수 내부 수정을 최소화하기 위해 상수 데이터(패널 너비, ajax URL)를 클래스 외부로 분리

Commit:
2.fetch API 변경

-carousel ajax를 XHR에서 fetch API로 변경

Commit:
3.검색자동완성 기능 구현

-대소문자 구분 없이 자동검색 키워드 하이라이트 표시

-검색창은 텍스트 입력에만 반응(유틸리티키 사용 금지)

-화살표 키, 엔터 키는 검색어 리스트가 완성됐을 때만 입력 가능

-화살표 키로 상하 이동시 무한 순환

-마우스 오버, 아웃시 배경색상 토글 및 클릭시 링크 이동

-입력 포커스 해제시 검색창 숨김
3.검색자동완성 기능 구현

-대소문자 구분 없이 자동검색 키워드 하이라이트 표시

-검색창은 텍스트 입력에만 반응(유틸리티키 사용 금지)

-화살표 키, 엔터 키는 검색어 리스트가 완성됐을 때만 입력 가능

-화살표 키로 상하 이동시 무한 순환

-마우스 오버, 아웃시 배경색상 토글 및 클릭시 링크 이동

-입력 포커스 해제시 검색창 숨김

crongro commented on amazon_cocoa.css:
```diff
@@ -55,8 +55,7 @@ table {
     background-color: rgb(34, 46, 62);
-    overflow: hidden;
-    color: white;
+    overflow: visible;
```
overflow 속성들에 대해서는 공부해볼만해요.

crongro commented on amazon_cocoa.html:
```diff
@@ -423,13 +454,13 @@ <h3 class="section-name">section-D</h3>
 <script type="text/javascript" src="plans_layer.js"></script>
 <script type="text/javascript" src="carousel_module.js"></script>
-
+<script type="text/javascript" src="auto_complete_module.js"></script>
```
script는 body태그 닫히기 직전에 주로 두어서, DOM 파싱작업을 방해하지 않죠. 여기에 두어도 됩니다.
async,defer속성은 또 무엇인지도 공부해두시고요. 이런것들은 초기로딩속도개선시 매우 중요한 방법들이에요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+// keywords AutoComplete Module.js step5
+
+const SETTINGVALUES = {
```
SETTINGVALUES 는 별도 파일로 분리해도 될 거 같네요.  ESMODULES를 사용한다면 import로 가져오면 되겠고요

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+// keywords AutoComplete Module.js step5
+
+const SETTINGVALUES = {
+    FETCHURL: "http://crong.codesquad.kr:8080/amazon/ac/",
```
FETCH_URL 처럼 대문자로 되어 있으면 구분을 (_)로 해주면 더 좋죠.
그리고 FETCHURL이라는 이름은 너무 추상적인 이름이니 더 구체적으로 지어주고요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    initEventListener() {
+        this.inputTextField.addEventListener("keyup", this.cancelFetch.bind(this));
```
이벤트 등록을 모아둔 거 잘하셨고요. arrow function을 사용하면 bind를 안해도 되는 거 기억하시고요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+        this.resultWindow.addEventListener("click", this.clickGoToLink.bind(this));
+    }
+
+    cancelFetch(evt) {
```
한줄짜리면 별도 함수로 만들어야하는가? 하는 생각이 좀 드네요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    triggerFetch(evt) {
+        if (this.inputTextField.value === "" && (evt.keyCode === 8 || evt.keyCode === 38 || evt.keyCode === 40)) return;
```
if(checkXXX()) return;

이런식으로 함수사용하기. 더 가독성있겠죠.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+        this.triggerTimeout = setTimeout(() => {
+            let composedURL = this.composeFetchURL();
+            this.operateFetch(composedURL);
+        }, 1000);
```
이런 숫자값1000 이 함수안에 있지 않고 함수밖으로 빼기. 심지어 class밖으로 빼서 class의 constructor에 넣어주는 방식으로 해도 됨.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    checkIsUtilityKey(evt) {
+        let result = this.KEYS.some(keys =>
```
간만에 보는 some 메서드네요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    operateFetch(URL) {
+        fetch(URL, { mode: "cors" })
```
promise패턴이 잘 이해되면, 
async/await 로도 변경해보세요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+            .then(response => response.json())
+            .then(this.checkCorrectKeyword.bind(this)) // 올바른 검색어인 경우 value, refTag 추출
+            .catch(error => console.log(error)) // 올바른 검색어가 아니면 에러 발생
+            .then(this.addItemLiElement.bind(this)) // 제품 링크 생성
```
catch 뒤에 then을 넣을 수 있는건가요? (몰라서 질문)

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+    addItemLiElement(itemsInfoList) {
+        let inputText = this.inputTextField.value.toLowerCase();
+        let itemLiElement = "";
+        itemsInfoList.forEach(value => {
```
아시겠지만 reduce를 쓸 수 있을거 같네요.
그리고 아래 li template이 좀 긴편인거 같긴한데요. 

이부분은 view를 처리하는 부분임으로 class에서 분리하는 것을 고민할 수도 있어요.
별도로 class에서 분리해서 처리하는 방법은 어떤방법을 써야할까요?

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+        if (this.suggestionItems === undefined) return; // 리스트가 완성되지 않았다면 작동하지 않는다. 
+        else if (this.suggestionItems.length === this.LIMIT && evt.keyCode === 40) {
+            this.moveDown(this.initialInputText);
+        } else if (this.suggestionItems.length === this.LIMIT && evt.keyCode === 38) {
```
this.suggestionItems.length === this.LIMIT 
이 비교문은 중복이네요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    // 키보드 아래로 이동
+    moveDown(initialInputText) {
```
moveDown 과 moveUp의 로직을 조금 더 개선해볼가치가 있네요.

- 함수를 더 하위함수로 분리해보기
- 두 함수간의 중복코드가 있다면 제거하기
- 1 이 의미하는 것을 구체화하기.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+    // 포커스가 벗어났을 때 검색창 닫기
+    closeWindowOnBlur(evt) {
+        let div = document.querySelector(".auto-complete-result-window");
+        if (window.getComputedStyle(div).borderColor === "rgb(1, 0, 0)") {
```
이렇게 비교하는 게 조금 위험해보여요.
borderColor는 언제든 바뀔 수 잇는 style 값 같아서요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    // 백스페이스로 모든 텍스트를 삭제했을 때 검색창 닫기
+    closeWindowOnInput(evt) {
```
아래위치한 메서드들은 명확하고 간결해서 좋네요.
evt를 파라미터로 받지 말고, destructuring을 활용해서 받으면 더 간결해질듯.

crongro reviewed (changes requested):
전체적으로 훌륭하게 구현했습니다.
UI 구현방식에 대해서 감이 좀 생기신 거 같네요.
고민의 흔적도 보여서 좋았어요.

리뷰내용 중 일부분을 수정해보고 가면 좋겠습니다.
모르는 것은 질문하시고요.

Commit:
4.프로퍼티 변경, ES모듈화, triggerFetch 수정

-SETTING_VALUES 프로퍼티 이름 수정
-SETTING_VALUES 외부 모듈 분리 및 export, import 처리
-cancelFetch 메소드를 triggerFetch 메소드 내에 통합

Commit:
commit 5

5.리팩토링, addItemLiElement 메소드 수정

-triggerFetch 메소드 수정
-moveUpDown  메소드 내 중복 삭제
-URL, li 템플릿 별도 모듈로 분리
-addItemLiElement 메소드 내 일부를 composeLiElement 메소드로 분리

Commit:
6.리팩토링, 기능추가, 성능개선

리팩토링
-operateFetch 메소드를 async/await 변경
-moveDown, moveUp 메소드 하위 메소드로 분리

기능추가
-입력한 키워드가 추천 검색어가 아닌 경우 별도 메시지 출력
-ESC키 입력시 결과창 닫기

성능개선
-디스트럭처링으로 일부 메소드 내 evt 객체 삭제

Commit:
7.리팩토링, 기능개선

리팩토링
-moveDown, moveUp 메소드 내 일부 숫자를 변수로 지정

기능개선
-closeWindowOnBlur 메소드에서 포커스 벗어날 시 dataset 속성 이용하여 개선
-추천 검색어 없음 메시지 출력시 불필요한 키보드 입력 금지

cocoa-salon commented:
피드백 사항을 반영하여 푸시하였습니다.  
(서버구동은 npm의 live-server 를 이용하였습니다.)
확인 부탁드립니다. 감사합니다.

crongro commented:
> 피드백 사항을 반영하여 푸시하였습니다.
> (서버구동은 npm의 live-server 를 이용하였습니다.)
> 확인 부탁드립니다. 감사합니다.

네 좋아요. 앞으로는 Express 서버를 적극 활용하시길!

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,253 @@
+        HIGHLIGHT_SPAN_TEMPLET,
+        NOT_FOUND_KEYWORD_MSG
+    }) {
+        this.URL = FETCH_REQUEST_URL_FOR_KEYWORDS;
```
막상 대문자가 많아지니 보기 어렵지 않나요?

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,253 @@
+        this.MIN_LIMIT = KEYWORDS_LIST_MIN_LIMIT;
+        this.TIMER = FETCH_REQUEST_TIMER;
+        this.KEYS = INPUT_PREVENTION_KEYS;
+        this.LI_TEMPLET = RESULT_LI_ELEMENTS_TEMPLET;
```
템플릿이면, template 로 써주세요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    // fetch 요청 url 구성
+    composeFetchRequestURL() {
```
유틸리티 역할을 하는 이런 메서드를 분리하는 것은 늘 좋죠.
나중에 다른 곳에서 재사용 될수도 있고요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,253 @@
+
+    // fetch api 호출
+    async operateFetch(URL) {
+        try {
```
에러처리가 상당히 중요한데 좋은 시도에요.
에러가 발생하는 경우를 확인까지 해서 catch안에 코드가 잘 동작하는지도 확인해보길 바랄게요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,253 @@
+    async operateFetch(URL) {
+        try {
+            const keywordsListText = await fetch(URL, { mode: "cors" });
+            const keywordsListObj = await keywordsListText.json();
```
async/await 는 현재 가장 일반적으로 사용하는 비동기연속처리 방법이에요. 잘 구현했고요.
비동기가 연속된 경우에 좀더 유용하긴 할거에요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,253 @@
+    // 항목 정보 배열 생성
+    generateItemsInfoList(jsonData) {
+        let itemsInfoList = jsonData.suggestions.reduce(function (list, value) {
+            return [...list, { value: value['value'], refTag: value['refTag'] }];
```
reduce를 사용하면, 완성된 하나의 결과를 만드는 과정임으로, 여기서는 immutable 하게 list를 만들지 않아도 될거 같긴하네요.
하지만 spread operator 도 써보고 잘하셨음.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,253 @@
+
+    // URL, li 엘리먼트 구성
+    composeLiElement(value, inputText) {
+        let regExp = new RegExp('\{\{\}\}');
```
오 template 파싱을 구현하셨군요. 템플릿라이브러리가 결국 하는 일이 이거긴해요.

이제부터넌, 
template literal를 사용해서 좀더 편리하게 templating작업을 해보세요.

crongro commented on auto_complete_module.js:
```diff
@@ -0,0 +1,253 @@
+    // 검색결과 클릭 시 해당 링크 이동
+    clickGoToLink(evt) {
+        evt.preventDefault();
+        if (evt.target.tagName === "LI" || evt.target.tagName === "A") {
```
if (target.tagName === "LI" || target.tagName === "A") { 
계속 반복되는 코드이니까.. 함수로 만들어도 될 듯.

crongro commented on setting_values.js:
```diff
@@ -0,0 +1,36 @@
+// setting_values.js step5
```
여러가지 다른 UI에 필요한 setting 정보도 여기들어갈 수 있으니, 
앞으로 다른 UI가 추가될때 여기 SETTING정보가 많아지면 어떻게 구분하고 재정리하지? 라는 고민을 해보세요.

crongro reviewed (approved):
네 수정을 하셔서 점점 좋아지고 있네요.
전체그림에서는 코드가 잘읽히고 작은 함수로 잘 만드셨어요.  👍 

리뷰남겨드리는 것들은 선택해서 수정하시고, 고민도 많이 하세요.
새롭게 배우는 것들이 많으니 궁금한 것들은 늘 질문하시고요.
